web/packages/teleterm/src/ui/services/workspacesService/accessRequestsService.ts

```diff
@@ -258,6 +329,7 @@ export type ResourceRequest =
       kind: 'kube';
       resource: {
         uri: KubeUri;
+        namespaces?: Map<ResourceUri, ResourceRequest>;
```

Comment:
Why does it need to be a map? Doesn't each namespace already have a `uri` field?

I see that Grzegorz used a set in his original suggestion. I assume this doesn't work because we don't operate on strings but rather whole objects which happen to have a URI field.

In that case, I'd opt for using a regular array. Using a map does make lookups and adding/deleting stuff easier. However, I'd say in this case it's more important to keep the data layer as simple as possible because performance is not the most important thing – I assume we're rarely going to see more than, say, 100 ResourceRequests.

Reply:
If a namespace doesn't need to be a part of `ResourceRequest`, then we could have a set of unique strings. But I haven't checked if that's possible, given that (as far as I understand) namespaces are supposed to act as individual items that you can request.

Reply:
I think we should not have the `namespace` type as part of the `ResourceRequest` union. In Connect's access requests world, a namespace is a part of `kube` resource, not the resource on its own. 

When it comes to the `addOrRemoveKubeNamespace`, I was thinking of using it directly in `useAccessRequsetCheckout`. Namespaces are not requested through `addOrRemoveResource` (like regular resources), but through `bulkToggleKubeResources`.

>If a namespace doesn't need to be a part of ResourceRequest, then we could have a set of unique strings

Yeah that's why I thought of `Set<KubeResourceNamespaceUri>` previously, since the URI keeps the same information as `PendingKubeResourceItem`. The only problem is that we have to encode/decode the URI from/to the shared type in `useAccessRequsetCheckout`. 
Now that I think of it, we don't even need the whole URI in `namespaces`, we could keep only a `namespaceId` since we could read `kubeId` from the `KubeUri`. But it doesn't matter that much.

>But I haven't checked if that's possible, given that (as far as I understand) namespaces are supposed to act as individual items that you can request.

I'm not sure if I understand, but I feel it is the opposite?
You can't request a namespace on its own, you need to select a kube cluster first and then select a namespace.

---------
I tried to turn these ideas into code, here's what I came up with https://github.com/gravitational/teleport/tree/gzdunek/teleterm-kube-namespace-refactor

Reply:
> > But I haven't checked if that's possible, given that (as far as I understand) namespaces are supposed to act as individual items that you can request.

> I'm not sure if I understand, but I feel it is the opposite?
You can't request a namespace on its own, you need to select a kube cluster first and then select a namespace.

i wasn't entirely sure how to translate question, both are correct?

in the web UI, user is required to select kube cluster first, then select its namespaces

in requests, yes, namespaces are individual requests

Reply:
The new implementation does solve the problem of namespaces being in two different places in the data model. 👍